server/src/main/java/com/google/coffeehouse/common/Book.java

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.coffeehouse.common;
+
+import com.google.coffeehouse.util.IdentifierGenerator;
+import com.google.coffeehouse.util.UuidWrapper;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Book encapsulates the information associated with each book being used by a Club.
+ * 
+ * <p>Implementes the {@link Saveable} interface because it will be able to save itself
+ * in the database.
+ */
+public class Book implements Saveable {
+  private String title;
+  private String author;
+  private String isbn;
+  private String bookId;
+
+  private Book(Builder builder) {
+    this.title = builder.title;
+    this.author = builder.author;
+    this.isbn = builder.isbn;
+    this.bookId = builder.idGenerator.generateId();
+  }
+
+  /** Overloaded static factory, calls other fromMap with null IdentifierGenerator. */
+  public static Book fromMap(Map bookInfo) {
+    return Book.fromMap(bookInfo, null);
+  }
+
+  /** 
+   * Creates a {@link Book} object from a Map with the relevant parameters as keys.
+   * @param bookInfo the Map that contains the information used to construct the Book. At a
+   *     minimum this includes a {@code "title"} key that is mapped to a String that describes
+   *     the title of the Book to be created. Optional keys include a {@code "author"} key 
+   *     that is mapped to a String describing the author of the book, as well as an 
+   *     {@code "isbn"} key that is mapped to a String that is the isbn number of the book
+   * @param idGen the {@link IdentifierGenerator} used when constructing the Book. If
+   *     null, the default generator will be used
+   * @return the created Book
+   * @throws IllegalArgumentException if no valid {@code "title"} key is defined
+   */
+  public static Book fromMap(Map bookInfo, IdentifierGenerator idGen) {
+    String title = (String) bookInfo.getOrDefault("title", null);
+    if (title == null) {
+      throw new IllegalArgumentException();
```

Comment:
Give this an error message so it's easier to debug exactly why the argument was illegal

Reply:
Very good point, this has been fixed and pushed,